akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala

```diff
@@ -93,17 +92,28 @@ private[io] final class AsyncDnsResolver(
           log.debug("{} cached {}", mode, resolved)
           sender() ! resolved
         case None =>
-          resolveWithResolvers(name, mode, resolvers)
-            .map { resolved =>
-              if (resolved.records.nonEmpty) {
-                val minTtl = (positiveCachePolicy +: resolved.records.map(_.ttl)).min
-                cache.put((name, mode), resolved, minTtl)
-              } else if (negativeCachePolicy != Never) cache.put((name, mode), resolved, negativeCachePolicy)
-              log.debug(s"{} resolved {}", mode, resolved)
-              resolved
+          val replyTo = sender()
+          resolveWithResolvers(name, mode, resolvers).onComplete {
+            _ match {
+              case Success(resolved) =>
+                if (resolved.records.nonEmpty) {
+                  val minTtl = (positiveCachePolicy +: resolved.records.map(_.ttl)).min
+                  cache.put((name, mode), resolved, minTtl) // thread-safe structure, safe in callback
+                } else if (negativeCachePolicy != Never) {
+                  cache.put((name, mode), resolved, negativeCachePolicy) // thread-safe structure, safe in callback
+                }
+
+                log.debug("{} resolved {}", mode, resolved)
+                replyTo ! resolved
+
+              case Failure(f) =>
+                replyTo ! Status.Failure(f)
             }
-            .pipeTo(sender())
+          }(ExecutionContexts.parasitic)
```

Comment:
somewhat scary to run that much code in parasitic, but as far as I can see nothing should throw there